File: lua_sol_test/src/properties.cpp
#include "properties.hpp"

namespace hd {

////////////////////////////////////////////////////////////////////////////////
// stream operators for properties
////////////////////////////////////////////////////////////////////////////////

std::ostream& operator<<(std::ostream& os, const size& s)
{
    os << s.width << ", " << s.height;
    return os;
}

std::ostream& operator<<(std::ostream& os, const col_rgb& c)
{
    os << c.r << ", " << c.g << ", " << c.b;
    return os;
}

std::ostream& operator<<(std::ostream& os, const position& p)
{
    os << p.x << ", " << p.y;
    return os;
}

std::ostream& operator<<(std::ostream& os, const speed& s)
{
    os << s.sx << ", " << s.sy;
    return os;
}

std::ostream& operator<<(std::ostream& os, const window& w)
{
    os << "size: " << w.s;
    return os;
}

std::ostream& operator<<(std::ostream& os, const font& f)
{
    os << "ffile: \"" << f.ffile << "\"" << std::endl;
    os << "fsize: " << f.fsize << std::endl;
    os << "col_rgb: " << f.col;
    return os;
}

std::ostream& operator<<(std::ostream& os, const circle& c)
{
    os << "name: \"" << c.name << "\"" << std::endl;
    os << "pos: " << c.pos << std::endl;
    os << "spd: " << c.spd << std::endl;
    os << "col_rgb: " << c.col << std::endl;
    os << "radius: " << c.radius;
    return os;
}

std::ostream& operator<<(std::ostream& os, const rectangle& r)
{
    os << "name: \"" << r.name << "\"" << std::endl;
    os << "pos: " << r.pos << std::endl;
    os << "spd: " << r.spd << std::endl;
    os << "col_rgb: " << r.col << std::endl;
    os << "size: " << r.s;
    return os;
}

} // namespace hd

File: lua_sol_test/input/properties.lua
--[[
    properties are given either as standalone lua type:

    type = {
        ...
    }

    or as indexed sequence of types with a naming convention
    for indexed sequences as type_seq:

    type_seq = {
        {...},
        {...}
    }
--]]
window = {
    size = { width = 600, height = 400 }
}

font = {
    ffile = "../fonts/arial.ttf",
    fsize = 24,
    col_rgb = { r = 255, g = 255, b = 255 }
}

circle_seq = {
    {
        name = "CGreen",
        position = { x = 100, y = 100 },
        speed = { sx = -0.03, sy = 0.02 },
        col_rgb = { r = 0, g = 255, b = 0 },
        radius = 50
    },
    {
        name = "CBlue",
        position = { x = 200, y = 200 },
        speed = { sx = 0.02, sy = 0.04 },
        col_rgb = { r = 0, g = 0, b = 255 },
        radius = 100
    }
}

rectangle_seq = {
    {
        name = "RRed",
        position = { x = 200, y = 200 },
        speed = { sx = 0.1, sy = 0.15 },
        col_rgb = { r = 255, g = 0, b = 0 },
        size = { width = 70, height = 25 }
    },
    {
        name = "RGreen",
        position = { x = 200, y = 200 },
        speed = { sx = 0.1, sy = -0.15 },
        col_rgb = { r = 0, g = 255, b = 0 },
        size = { width = 30, height = 45 }
    }
}
